File: app/src/main/java/com/example/sleppify/AnimatedEqualizerView.java
package com.example.sleppify;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.View;
import androidx.annotation.Nullable;

import java.util.Random;

public class AnimatedEqualizerView extends View {

    private final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final float[] barHeights = {0.2f, 0.2f, 0.2f};
    private final float[] targetHeights = {0.2f, 0.2f, 0.2f};
    private final Random random = new Random();
    private boolean animating = false;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private int barColor = 0xFFFFFFFF;

    private final Runnable animationRunnable = new Runnable() {
        @Override
        public void run() {
            if (!animating) return;

            boolean changed = false;
            for (int i = 0; i < 3; i++) {
                if (Math.abs(barHeights[i] - targetHeights[i]) < 0.05f) {
                    targetHeights[i] = 0.3f + random.nextFloat() * 0.7f;
                }
                
                if (barHeights[i] < targetHeights[i]) {
                    barHeights[i] += 0.1f;
                } else {
                    barHeights[i] -= 0.1f;
                }
                changed = true;
            }

            if (changed) {
                invalidate();
            }
            handler.postDelayed(this, 30);
        }
    };

    public AnimatedEqualizerView(Context context) {
        super(context);
        init(null);
    }

    public AnimatedEqualizerView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public AnimatedEqualizerView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    private void init(@Nullable AttributeSet attrs) {
        if (attrs != null) {
            TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.AnimatedEqualizerView);
            barColor = a.getColor(R.styleable.AnimatedEqualizerView_barColor, 0xFFFFFFFF);
            a.recycle();
        }
        paint.setColor(barColor);
        paint.setStyle(Paint.Style.FILL);
    }

    public void setBarColor(int color) {
        this.barColor = color;
        paint.setColor(color);
        invalidate();
    }

    public void setAnimating(boolean animating) {
        if (this.animating == animating) return;
        this.animating = animating;
        if (animating) {
            handler.post(animationRunnable);
        } else {
            handler.removeCallbacks(animationRunnable);
            // Reset to low height
            for (int i = 0; i < 3; i++) {
                barHeights[i] = 0.2f;
                targetHeights[i] = 0.2f;
            }
            invalidate();
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        float width = getWidth();
        float height = getHeight();
        float barWidth = width / 5f;
        float spacing = width / 10f;

        for (int i = 0; i < 3; i++) {
            float barHeight = height * barHeights[i];
            float left = spacing + i * (barWidth + spacing);
            float top = height - barHeight;
            float right = left + barWidth;
            float bottom = height;
            
            canvas.drawRect(left, top, right, bottom, paint);
        }
    }
}

File: app/src/main/java/com/example/sleppify/AnimatedEqualizerView.kt
package com.example.sleppify

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.View

class AnimatedEqualizerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val barHeights = floatArrayOf(0.2f, 0.2f, 0.2f)
    private val targetHeights = floatArrayOf(0.2f, 0.2f, 0.2f)

    // Deterministic keyframe pattern — same table for all bars, offset by phase
    private val keyframes = floatArrayOf(0.3f, 0.75f, 0.45f, 0.9f, 0.55f, 0.35f, 0.8f, 0.5f)
    private val phaseIndex = intArrayOf(0, 3, 5) // staggered start phases per bar

    private var animating = false
    private val handler = Handler(Looper.getMainLooper())
    private var barColor = -0x1

    private val animationRunnable = object : Runnable {
        override fun run() {
            if (!animating) return

            var changed = false
            for (i in 0..2) {
                if (Math.abs(barHeights[i] - targetHeights[i]) < 0.05f) {
                    phaseIndex[i] = (phaseIndex[i] + 1) % keyframes.size
                    targetHeights[i] = keyframes[phaseIndex[i]]
                }

                if (barHeights[i] < targetHeights[i]) {
                    barHeights[i] += 0.04f
                } else {
                    barHeights[i] -= 0.04f
                }
                changed = true
            }

            if (changed) {
                invalidate()
            }
            handler.postDelayed(this, 60)
        }
    }

    init {
        attrs?.let {
            val a = context.obtainStyledAttributes(it, R.styleable.AnimatedEqualizerView)
            barColor = a.getColor(R.styleable.AnimatedEqualizerView_barColor, -0x1)
            a.recycle()
        }
        paint.color = barColor
        paint.style = Paint.Style.FILL
    }

    fun setBarColor(color: Int) {
        this.barColor = color
        paint.color = color
        invalidate()
    }

    fun setAnimating(animating: Boolean) {
        if (this.animating == animating) return
        this.animating = animating
        if (animating) {
            handler.post(animationRunnable)
        } else {
            handler.removeCallbacks(animationRunnable)
            // Reset to low height
            for (i in 0..2) {
                barHeights[i] = 0.2f
                targetHeights[i] = 0.2f
                phaseIndex[i] = i * 3 // reset staggered phases
            }
            invalidate()
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val width = width.toFloat()
        val height = height.toFloat()
        val barWidth = width / 5f
        val spacing = width / 10f

        for (i in 0..2) {
            val barHeight = height * barHeights[i]
            val left = spacing + i * (barWidth + spacing)
            val top = height - barHeight
            val right = left + barWidth
            val bottom = height

            canvas.drawRect(left, top, right, bottom, paint)
        }
    }
}
